### ISSUE TITLE
byzcoin cannot store new skipblocks when skipchain service is linked

### ISSUE THREAD

**Issue by jeffallen** [dedis/cothority#1756]
Once you have run `scmgr link add private.toml`, further calls to StoreSkipblock will fail, even if they are coming internally (i.e. from sister service Byzcoin on the same conode).

The workaround is to link to the conode while you are doing what you need to do, then unlink again (`scmgr link del $host:$port`).

We need to find a more workable solution. One possibility would be to have a setter on type StoreSkipblock, making it possible to set a private field named `internal`. Callers on the network would never be able to set that, but sister services would be able to.

A more general solution would be for onet to be able to signal somehow, that requests are originating locally instead of over the network. This would apply to all services, and then Skipchain could use this to decide to allow unsigned block writes from internal services.

**Comment by ineiti** [comment]
Would you consider the following too hacky?

```
func (s *Service) StoreSkipBlock(psbd *StoreSkipBlock) (*StoreSkipBlockReply, error) {
    // check for s.Storage.Clients
    return s.storeSkipBlockInternal(psbd)
}

func (s *Service) storeSkipBlockInternal(psbd *StoreSkipBlock) (*StoreSkipBlockReply, error) {
    // Do the rest
}
```

Then in byzcoin, you could call `storeSkipBlockInternal`, which is not accessible from the outside.

**Comment by jeffallen** [comment]
That would not work, because of Go namespace rules. But something like it could work, where you wrap the *StoreSkipBlock up into another type, which was not registered for use across onet.

As for too hacky, yes, pretty hacky. And probably not backwards compatible, which we need now that v3 is released.

Still thinking.

**Comment by ineiti** [comment]
But if you make `storeSkipBlockInternal` uppercase, it would work, wouldn't it? And it would be backward-compatible, as the API would not change, you only add a new API method.

**Comment by jeffallen** [comment]
Oops, yes. I was forgetting that you can have a public method which is not registered (somehow I was thinking it was the type that is registered, not the method). So the hacky way is available to us, but I will also experiment with a more general solution in Onet.

**Comment by jeffallen** [comment]
Here's one time I managed to reproduce it:
```
$ ./run_nodes.sh -n 5 -c -t -v 2

Other window:

  export BC_CONFIG=~/work/cothority/conode/
  bcadmin create -roster public.toml -i 1s
  export BC=bc-4a594ac283111d952299d513f01710811cdc927213afba89693b87e7ad2dd756.cfg 
  bcadmin latest
  bcadmin darc show
  bcadmin darc rule -update -rule spawn:xxx -identity ed25519:f5c918da785354cf5ac3120ece5554f3e8e85e543136ae5e0a9588c747386ae6 
  # ^ makes new block
  scmgr link a co1/private.toml 
  scmgr link a co2/private.toml 
  scmgr link a co3/private.toml 
  scmgr link a co4/private.toml 
  scmgr link a co5/private.toml 
  # try to make a new block:
  bcadmin darc rule -replace -rule spawn:xxx -identity ed25519:f5c918da785354cf5ac3120ece5554f3e8e85e543136ae5e0a9588c747386ae6 
```

The error from the leader was:
```
E 19/11/04 11:34:38.230633000: (byzcoin.(*txPipeline).processTxs.func2.1:  374) - failed to propose block:  cannot create new skipblock without authentication
E 19/11/04 11:34:38.230747000: (  byzcoin.(*txPipeline).processTxs.func2:  385) - reverting to last known state because proposal refused: cannot create new skipblock without authentication
E 19/11/04 11:34:57.036051000: (                  v3.wsHandler.ServeHTTP:  236) - Got an error while executing ByzCoin/AddTxRequest: transaction didn't get included after 20s (2 * t_block * 10)
```

**Comment by nkcr** [comment]
Fixed in #1810
